src/compiler/checker.ts

```diff
@@ -34268,7 +34273,39 @@ namespace ts {
                 const saveCurrentNode = currentNode;
                 currentNode = node;
                 instantiationCount = 0;
+
+                const oldTypeCount = typeCount;
+                const oldSymbolCount = symbolCount;
+
                 checkSourceElementWorker(node);
+
+                if (node.kind >= SyntaxKind.FirstStatement && node.kind <= SyntaxKind.LastStatement) {
+                    if (ignoreExpensiveStatement) {
+                        // The first statement is unfairly blamed for a lot of lib types
+                        ignoreExpensiveStatement = false;
+                    }
+                    else {
+                        const typeDelta = typeCount - oldTypeCount;
+                        const symbolDelta = symbolCount - oldSymbolCount;
+                        const record = { node, typeDelta, symbolDelta };
+
+                        let i = 0;
+                        for (const record of expensiveStatements) {
+                            if (record.typeDelta < typeDelta) {

```

Comment:
In teamroom discussions, there has seemed to be more interest in type count than symbol count, so I used that for sorting.

Reply:
I wonder if it'll be able to detect groups of statements that are cheap individually but are executed so often that they are actually expensive.

eg. I have some type `ConvertNumbersToStringsInInterface<I>` that will be not so expensive as individual call.

But let's say I use it on `ViewStyle` interface which can be included 100s of properties.

And then let's say I use it in every view I've got in my app. Then I can end up having 1000s of relatively cheap operations instead of 1 huge one.

Maybe it'll be possible to track the total time of each type being 'part of' for other operations?

Reply:
@pie6k I had an earlier version that listed top calls by both individual invocation and aggregate, but it proved to be too difficult to read the output.  Anecdotally, this unsophisticated heuristic has been helpful in some real world projects, but, as you point out, it has numerous limitations.